tests/ptf/fabric_test.py

```diff
@@ -698,6 +751,172 @@ def runIPv4UnicastTest(self, pkt, next_hop_mac,
         testutils.verify_no_other_packets(self)
 
 
+class DoubleVlanTerminationTest(FabricTest):
+
+    def runRouteAndPushTest(self, pkt, next_hop_mac,
+                            prefix_len=24,
+                            exp_pkt=None,
+                            next_id=None,
+                            next_vlan_id=None,
+                            next_inner_vlan_id=None,
+                            in_tagged=False,
+                            dst_ipv4=None,
+                            routed_eth_types=(ETH_TYPE_IPV4,),
+                            verify_pkt=True):
+        """
+        Route and Push test case. The switch output port is expected to send double tagged packets.
+        The switch routes the packet to the correct destination and adds the double VLAN tag to it.
+        :param pkt:
+        :param next_hop_mac:
+        :param prefix_len:
+        :param exp_pkt:
+        :param next_id:
+        :param next_vlan_id:
+        :param next_inner_vlan_id:
+        :param in_tagged:
+        :param dst_ipv4:
+        :param routed_eth_types:
+        :param verify_pkt:
+        :return:
+        """
+
+        if IP not in pkt or Ether not in pkt:
+            self.fail("Cannot do IPv4 test with packet that is not IP")
+
+        pkt_is_tagged = False
+        if Dot1Q in pkt:
+            in_vlan = pkt[Dot1Q].vlan
+            in_tagged = True
+            pkt_is_tagged = True
+        else:
+            in_vlan = VLAN_ID_1
+
+        next_id = 100 if next_id is None else next_id
+
+        if dst_ipv4 is None:
+            dst_ipv4 = pkt[IP].dst
+        switch_mac = pkt[Ether].dst
+
+        # Setup port 1
+        self.setup_port(self.port1, vlan_id=in_vlan, tagged=in_tagged)
+        # Setup port 2: packets on this port are double tagged packets
+        self.setup_port(self.port2, vlan_id=next_vlan_id, double_tagged=True, inner_vlan_id=next_inner_vlan_id)
+
+        # Forwarding type -> routing v4
+        for eth_type in routed_eth_types:
+            self.set_forwarding_type(self.port1, switch_mac, eth_type,
+                                     FORWARDING_TYPE_UNICAST_IPV4)
+
+        # Routing entry.
+        self.add_forwarding_routing_v4_entry(dst_ipv4, prefix_len, next_id)
+        self.add_next_routing(next_id, self.port2, switch_mac, next_hop_mac)
+
+        # Push double vlan
+        self.add_next_double_vlan(next_id, next_vlan_id, next_inner_vlan_id)
+
+        if exp_pkt is None:
+            # Build exp pkt using the input one.
+            exp_pkt = pkt.copy()
+            exp_pkt = pkt_add_vlan(exp_pkt, next_inner_vlan_id)
+            exp_pkt = pkt_add_vlan(exp_pkt, next_vlan_id)
+            exp_pkt = pkt_route(exp_pkt, next_hop_mac)
+            if in_tagged:
+                exp_pkt = pkt_remove_vlan(exp_pkt, in_vlan)
+
+        if in_tagged and not pkt_is_tagged:
+            pkt = pkt_add_vlan(pkt, vlan_vid=in_vlan)
+
+        testutils.send_packet(self, self.port1, str(pkt))
```

Comment:
The pattern for all test cases is to call `testutils.send_packet` after building the exp_pkt